tls/s2n_handshake_io.c

```diff
@@ -1529,6 +1530,12 @@ static int s2n_handshake_read_io(struct s2n_connection *conn)
             continue;
         }
 
+        /* Check for missing Certificate Requests to surface a more specific error */
+        if (EXPECTED_MESSAGE_TYPE(conn) == TLS_CERT_REQ) {
+            POSIX_ENSURE(message_type == TLS_CERT_REQ,
+                    S2N_ERR_MISSING_CERT_REQUEST);
+        }
+
         POSIX_ENSURE(record_type == EXPECTED_RECORD_TYPE(conn), S2N_ERR_BAD_MESSAGE);
         POSIX_ENSURE(message_type == EXPECTED_MESSAGE_TYPE(conn), S2N_ERR_BAD_MESSAGE);
```

Comment:
My first draft of this just replaced the error in the existing check with a method that choose an error type based on the EXPECT_MESSAGE_TYPE, instead of adding a new check:
```suggestion
        POSIX_ENSURE(record_type == EXPECTED_RECORD_TYPE(conn), S2N_ERR_BAD_MESSAGE);
        POSIX_ENSURE(message_type == EXPECTED_MESSAGE_TYPE(conn), s2n_expected_message_errno(conn));
```
Unfortunately, we have several grep_simple_mistakes checks that interfered. The biggest issue was that we require POSIX_ENSURE to take an "S2N_ERR" argument. I could work around that with a macro that started with S2N_ERR, but at that point it seemed like more of a hack than just adding the new check :/ And I don't want to modify the grep_simple_mistakes check because honestly it's a good check.

Reply:
If reviewers prefer the old way, I could probably put a hacky exception in grep_simple_mistakes. Like `grep -v "s2n_handshake_io.c.*s2n_expected_message_errno` the results.

Reply:
You could potentially use a local define... Not sure if it's any better, though

```c
#define S2N_ERR_EXPECTED_ERRNO s2n_expected_message_errno(conn)
POSIX_ENSURE(message_type == EXPECTED_MESSAGE_TYPE(conn), S2N_ERR_EXPECTED_ERRNO);
#undef S2N_ERR_EXPECTED_ERRNO
```